### DISCUSSION TITLE
RAID

### DISCUSSION THREAD

**Discussion by 3Juhwan** [2024-woowacourse-study/level-interview#267] [Computer Architecture]
.

**Comment by le2sky** [comment]
**RAID(Redundant Array of Independent Disks)** 는 수 많은 데이터들을 안전하게 저장하거나 성능을 높이기 위해 여러 하드 디스크나 SSD를 마치 하나의 장치처럼 사용하는 기술이다.

**RAID 구성 방법(RAID 레벨이라고도 함)**

**RAID-0**은 여러 개의 보조기억 장치에 데이터를 단순히 나누어 저장하는 구성 방식이다.  특정 데이터를 저장할 때 하드 디스크는 각 장치에 번갈아 데이터를 저장한다. 이때 줄무늬처럼 분산되어 저장된 데이터를 스트라이프라고 하며, 분산하여 저장하는 방식을 스트라이핑이라고 한다.

- 데이터를 분산하여 저장하기 때문에 저장된 데이터를 읽고 쓰는 속도가 빨라진다.
- 하나의 대용량 저장 장치를 사용하면 여러 번에 걸쳐 읽고 쓰는 데이터를 동시에 읽고 쓸 수 있기 때문이다.
- ex) 4TB 저장 장치 한개 읽기 속도 < RAID 0 구성으로된 1TB 장치 네 개 읽기 속도
- (단점) 저장된 정보가 안전하지 않다는 단점이 존재한다. (디스크 하나 부서지면..)

--- 

**RAID-1**은 완전한 복사본을 만드는 구성이며 미러링이라고도 부른다. 

- 디스크에 문제가 발생해도 복구가 가능하다.
- (단점) 하드 디스크 개수가 한정되었을 때 사용 가능한 용량이 적어진다.
- (단점) RAID-0에 비해 쓰기 속도가 느리다.(두 곳에 써야 하므로)

---

**RAID-4**는 복사본을 만드는 대신 오류를 검출하고 복구하기 위한 정보(패리티 비트)를 저장한 장치를 두는 구성 방식이다.  

- RAID-1 보다 적은 하드 디스크로도 데이터를 안전하게 보관할 수 있다.
- (단점) 새로운 데이터가 저장될 때마다 패리티를 저장하는 디스크에도 데이터를 써야 하므로 병목 현상이 발생할 수 있다.

---

**RAID-5**는 RAID-4의 병목 현상을 극복하기 위해 패리티 정보를 분산하여 저장한다. 

<p align = "center">
    <img height = 300 width = 500 src = "https://github.com/user-attachments/assets/f10b6809-89e5-45dd-bad7-da60ebc960b5" />
</p>

---

**RAID-6**는 서로 다른 두 개의 패리티를 두는 방식이다. 

<p align = "center">
    <img height = 300 width = 500 src = "https://github.com/user-attachments/assets/e4e17549-5ad1-4d9c-b369-b1bae534e6f0" />
</p>

- 오류 검출 및 복구 수단이 두 개가 생긴다.
- RAID-4나 RAID-5 보다 안전한 구성이다.
- (단점) 패리티가 두 개이므로 쓰기 속도는 RAID-5보다 느리다.

**Reply by le2sky** [reply]
++ 궁금해서 찾아본거 메모

Amazon RDS DB 인스턴스는 데이터베이스 및 로그 스토리지에 Amazon Elastic Block Store(EBS) 볼륨을 사용합니다. RDS(RDS for SQL Server 제외)는 필요한 스토리지 용량에 따라 자동으로 데이터를 여러 Amazon EBS 볼륨에 스트라이핑하여 성능을 강화합니다. SSD 스토리지 유형이 있는 RDS DB 인스턴스는 RAID 0 구성에서 1개 또는 4개의 스트라이프 Amazon EBS 볼륨으로 지원됩니다. 설계상 RDS DB 인스턴스의 스토리지 수정 작업은 진행 중인 데이터베이스 작업에 미치는 영향을 최소화합니다.

출처 : https://docs.aws.amazon.com/ko_kr/AmazonRDS/latest/UserGuide/USER_PIOPS.StorageTypes.html#USER_PIOPS.IOIntensive
